### PR TITLE
Added typescript support.

### PR DESCRIPTION
**Pull request by boleerandersen**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -12,6 +12,7 @@ By default the following directory structure would end up something like the fol
 
     |---- bower_components
     |---- dist
+    |---- tmp
```
Comment (langecode) [inline README.md]: Why would you want to introduce a `tmp` directory? Currently all generated sources are kept under `dist`I would recommend still keeping all generated code here, e.g., output from the transpiling in dist/ts (or of course `options.dist` concatenated with `ts`).

File: src/build.gulp.js [hunk 2]
```diff
@@ -2,13 +2,12 @@
...
 var angularFilesort = require('gulp-angular-filesort'),
     concat = require('gulp-concat'),
-    gulp = require('gulp'),
     ngTemplateStrings = require('gulp-ng-template-strings'),
     path = require('path');
 
-module.exports = function (options) {
-    gulp.task('templates', function () {
-        return gulp.src([path.join(options.src, '**/*.js'), '!**/*.spec.js'])
+module.exports = function (options, gulp) {
```
Comment (langecode) [inline src/build.gulp.js]: Now why would you want to parse `gulp` around when the package system is capable of taking care of this?

File: src/ts.gulp.js [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+     */
+    gulp.task('ts', function () {
+        var tsProject = ts.createProject({
+                compilerOptions: defaults
+            }),
+            tsResult = gulp.src(path.join(options.src, '**/*.ts'))
+                .pipe(sourcemaps.init())
+                .pipe(ts(tsProject));
+
+        return tsResult.js
```
Comment (langecode) [inline src/ts.gulp.js]: I expect that you are totally certain that this works with the parallelization in Gulp? I think it looks a little bit suspect that the pipes are done in two goes (and we have seen a lot of very non-deterministic problems when parallelization kicks in) - but is might be ok depending on how the inner workings is here.

File: src/ts.gulp.js [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+            .pipe(sourcemaps.write('maps'))
+            .pipe(gulp.dest(options.tmp));
+    });
+    
+};
+
+
+
```
Comment (langecode) [inline src/ts.gulp.js]: Please remove these extra line feeds. :-)

File: src/index.js [hunk 5]
```diff
@@ -1,21 +1,22 @@
 'use strict';
 
-var assign = require('object-assign'),
-    gulp = require('gulp');
+var assign = require('object-assign')
 
-module.exports = function (options) {
+module.exports = function (options, gulp) {
```
Comment (langecode) [inline src/index.js]: Why add `gulp`? To me it just seems to complicate things.

**Comment by boleerandersen** [comment]
Yes, one would think that the node.js module system would make this unneccesary. But if gulp isn't passed, a new instance of gulp is created in each require in each submodule, resulting in no gulp tasks being defined (visible in the top module).

I don't now if this is a particully issue with our node version. Gript is doing it in the same way, maybe due to having experinced the same issue.

File: src/build.gulp.js (hunk 2 above)
Comment (boleerandersen) [inline src/build.gulp.js]: Yes, one would think that the node.js module system would make this unneccesary. But if gulp isn't passed, a new instance of gulp is created in each require in each submodule, resulting in no gulp tasks being defined (visible in the top module).

I don't now if this is a particully issue with our node version. Gript is doing it in the same way, maybe due to having experinced the same issue.

Comment (langecode) [inline src/build.gulp.js]: But the package as-is worked - so why shoudn't it work when ts is added?

File: src/test.gulp.js [hunk 6]
```diff
@@ -1,29 +1,29 @@
...
             dependencies: true,
             devDependencies: true
         });
         new Server({
             configFile: __dirname + '/karma.conf.js',
-            files: bowerDeps.js.concat(path.join(options.src, '**/*.js')),
+            files: bowerDeps.js.concat(path.join(options.src, '**/*.js'), path.join(options.ts, '/**/*.js')),
             singleRun: true
-        }, cb).start();
+        }, function() {cb();}).start();
```
Comment (langecode) [inline src/test.gulp.js]: Isn't this the same? Passing the `cb` function directly or creating an anonymous function that calls the `cb` function? Unless I am mistaken it is exactly the same thing...

File: src/build.gulp.js [hunk 7]
```diff
@@ -7,8 +7,8 @@ var angularFilesort = require('gulp-angular-filesort'),
     path = require('path');
 
 module.exports = function (options) {
-    gulp.task('templates', function () {
-        return gulp.src([path.join(options.src, '**/*.js'), '!**/*.spec.js'])
+    gulp.task('templates', ['ts'], function () {
+        return gulp.src([path.join(options.src, '**/*.js'), path.join(options.ts, '/**/*.js'), '!**/*.spec.js'])
```
Comment (langecode) [inline src/build.gulp.js]: This is to make sure templates referenced from typescript code is included? Makes sense :)

File: src/build.gulp.js [hunk 8]
```diff
@@ -25,5 +25,5 @@ module.exports = function (options) {
             .pipe(gulp.dest(path.join(options.dist, 'build')));
     });
 
-    gulp.task('build', ['lint', 'test', 'concat', 'html']);
+    gulp.task('build', ['ts', 'lint', 'test', 'concat', 'html']);
```
Comment (langecode) [inline src/build.gulp.js]: Well, both `test` and `templates` are already dependent on `ts` and I guess those are really the tasks which should be dependent. No big deal, but I would probably not have it here.

**Comment by langecode** [comment]
I think it looks good. Two small comments:

- I wouldn't have the `build` task be directly dependent on the `ts` task - you have already created the necessary dependencies in `test` and `templates` and not having it in `build` will allow gulp to do a more optimal run.
- I cannot see the reason to wrap the cb function in another function? It seems just an unnecessary indirection as no extra functionality is provided.

But I think those are very small issues - so I think we are about ready to merge it in (and probably do a new release to npm).

**Pull request merged**